src/zap.c

```diff
@@ -5206,27 +5245,73 @@ int osym, dmgtyp;
+ * player */
+int
+destroy_items(mon, dmgtyp, dmg_in)
+struct monst *mon; /* monster whose invent is being subjected to destruction */
+int dmgtyp,  /* AD_**** - currently only cold, fire, elec */
+    dmg_in;  /* the amount of HP damage the attack dealt */
 {
     register struct obj *obj;
-    int i, deferral_indx = 0;
-    /* 1+52+1: try to handle a full inventory; it doesn't matter if
-      inventory actually has more, even if everything should be deferred */
-    unsigned short deferrals[1 + 52 + 1]; /* +1: gold, overflow */
+    int i, defer;
+    int limit; /* max amount of item stacks destroyed, based on damage */
+    struct {
+        unsigned oid;
+        boolean deferred;
+    } items_to_destroy[MAX_ITEMS_DESTROYED];
+    int elig_stacks = 0; /* number of destroyable objects found so far */
+    boolean u_carry = (mon == &g.youmonst);
+    /* this is a struct obj** because we might destroy the first item in it */
+    struct obj **objchn = u_carry ? &g.invent : &mon->minvent;
+    int dmg_out = 0; /* damage caused by items getting destroyed */
 
-    (void) memset((genericptr_t) deferrals, 0, sizeof deferrals);
-    /*
-     * Sometimes destroying an item can change inventory aside from
+    /* special effect of extrinsic resistances: they protect all items from
+     * their respective damage types */
+    if ((dmgtyp == AD_FIRE
+         && (u_carry ? EFire_resistance : (mon->mextrinsics & MR_FIRE)))
+        || (dmgtyp == AD_COLD
+            && (u_carry ? ECold_resistance : (mon->mextrinsics & MR_COLD)))
+        || (dmgtyp == AD_ELEC
+            && (u_carry ? EShock_resistance : (mon->mextrinsics & MR_ELEC)))) {
+        return 0;
+    }
+
+    /* initialize items_to_destroy */
+    for (i = 0; i < MAX_ITEMS_DESTROYED; ++i) {
+        /* 0 should not be a valid o_id for anything */
+        items_to_destroy[i].oid = 0;
+        items_to_destroy[i].deferred = FALSE;
+    }
+
+    /* Don't straight up destroy all items with an equal chance; limit it based
+     * on the amount of damage being dealt by the source of the item
+     * destruction. */
+    limit = dmg_in / DMG_DESTROY_SCALE;
+    if (dmg_in % DMG_DESTROY_SCALE > rn2(DMG_DESTROY_SCALE)) {
+        limit++; /* dmg = 9: 87.5% chance limit = 1, 12.5% limit = 2, etc */
+    }
+    if (limit > MAX_ITEMS_DESTROYED) {
+        /* in case of incredibly high damage, prevent from overflowing
+         * items_to_destroy */
+        limit = MAX_ITEMS_DESTROYED;
+    }
+
+    /* Sometimes destroying an item can change inventory aside from
      * the item itself (cited case was a potion of unholy water; when
      * boiled, potionbreathe() caused hero to transform into were-beast
      * form and that resulted in dropping or destroying some worn armor).
      *
-     * Unlike other uses of the object bybass mechanism, destroy_item()
+     * Unlike other uses of the object bypass mechanism, destroy_items()
```

Comment:
Should change it back to bybass, you will lose your Bass Pro Shop sponsorship otherwise.

Reply:
Those jerks never did anything for me, I'm cutting ties.